tensorflow/core/kernels/bincount_op.cc

```diff
@@ -75,34 +67,86 @@ class BincountOp : public OpKernel {
           for (int64 i = start_ind; i < limit_ind; i++) {
             int32 value = arr(i);
             if (value < size) {
-              if (has_weights) {
-                partial_bins(worker_id, value) += weights(i);
-              } else {
-                // Complex numbers don't support "++".
-                partial_bins(worker_id, value) += T(1);
-              }
+              // Complex numbers don't support "++".
+              partial_bins(worker_id, value) += T(1);
             }
           }
         });
-    TensorShape output_shape({size});
-    Tensor* output_t;
-    OP_REQUIRES_OK(ctx, ctx->allocate_output(0, output_shape, &output_t));
+
     // Sum the partial bins along the 0th axis.
     Eigen::array<int, 1> reduce_dims({0});
-    output_t->flat<T>().device(ctx->eigen_cpu_device()) =
-        partial_bins.sum(reduce_dims);
+    output.device(context->eigen_cpu_device()) = partial_bins.sum(reduce_dims);
+    return Status::OK();
   }
 };
 
-#define REGISTER(TYPE)                                               \
+}  // namespace functor
+
+template <typename Device, typename T>
+class BincountOp : public OpKernel {
+ public:
+  explicit BincountOp(OpKernelConstruction* ctx) : OpKernel(ctx) {}
+
+  void Compute(OpKernelContext* ctx) override {
+    const Tensor& arr_t = ctx->input(0);
+    const Tensor& size_tensor = ctx->input(1);
+    const Tensor& weights_t = ctx->input(2);
+
+    int32 size = size_tensor.scalar<int32>()();
+    OP_REQUIRES(ctx, size >= 0, errors::InvalidArgument(
+                                    "size (", size, ") must be non-negative"));
+
+    const bool has_weights = weights_t.NumElements() > 0;
+    OP_REQUIRES(ctx, !(has_weights && arr_t.shape() != weights_t.shape()),
+                errors::InvalidArgument(
+                    "If weights are passed, they must have the same shape (" +
+                    weights_t.shape().DebugString() + ") as arr (" +
+                    arr_t.shape().DebugString() + ")"));
+    const auto arr = arr_t.flat<int32>();
+    const auto weights = weights_t.flat<T>();
+
+    if (weights.size() == 0) {
+      Tensor* output_t;
+      OP_REQUIRES_OK(ctx,
+                     ctx->allocate_output(0, TensorShape({size}), &output_t));
+      auto output = output_t->flat<T>();
+      OP_REQUIRES_OK(
+          ctx, functor::BincountFunctor<Device, T>::Compute(ctx, arr, output));
+    } else {
```

Comment:
Please handle the case of weights being None in the graph (ie edit the python code for the bincount op) by calling unsorted_segment_sum there after reduce_max.  The C++ code for bincount only needs to handle the case where the weights are implicitly one.

It's better to not have high level ops alias to the same underlying calculation; it only makes analyzing the graph harder.

Reply:
Thanks @ekelsen. That part has been moved to python side.